### PR TITLE
[ENH] image gradient consistency checking

### PR DESCRIPTION
**Pull request by dPys**
Builds upon #26 by adding a check that gradient encoding patterns found in the b-values correspond to those found in signal intensity variance across 4d volumes of the dwi image.

Disclaimer: this is experimental (sklearn's MeanShift is magic), but potentially invaluable. I still need to create a test or two, but figured I'd proceed to at least open the PR. So far, this function verifiably works on all different kinds of images that I've tried it with-- including multishell, dki, etc. 

If the check fails for any reason, a simple userwarning is raised (i.e. rather than crashing the whole workflow) so as to minimally alert the user to look at their images and ensure bval correspondence. This tool is also valuable for any changes made to the image volumes (e.g. removal of high-motion or FD ones), where an accompanying update to the gradient table becomes necessary. This check could be re-run after any such update to ensure consistency is maintained (it runs in ~1-2 seconds provided that the dwi image dataobject is already loaded into memory)...

@dPys

### CODE REVIEW

File: dmriprep/utils/vectors.py [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+        ix_vec = np.array(ix_vec)
+
+        if n_clusters_ != len(np.unique(gtab.bvals)):
+            raise UserWarning('Number of unique b-values does not correspond to number of unique signal gradient '
+                              'encoding intensities in the dwi image.')
+
+        if np.all(np.isclose(ix_img, ix_vec)) is True:
+            raise UserWarning('Positions of b-value B0\'s and shell(s) do not correspond to signal intensity '
+                              'fluctuation patterns in the dwi image.')
+    return
```
Comment (oesteban) [inline dmriprep/utils/vectors.py]: Is this the core of this PR?

Do you think it is worth waiting on #26 to be merged and then add this function to it?

Comment (dPys) [reply dmriprep/utils/vectors.py]: Yes, this is the PR. Will rebase this addition with the latest merge from #26.

**Comment by dPys** [comment]
@oesteban  - rebased to latest master merges, test added. Boolean to activate for CheckGradientTable interface.

**Comment by oesteban** [comment]
Can we close this one?

**Comment by dPys** [comment]
This one we might actually want to keep open since it adds an experimental feature unrelated to emc and not quite ready for primetime integration into the vectors interface. Could be useful, but would need tests. Can discuss whether we consider it a valuable target tomorrow when we meet?

**Comment by oesteban** [comment]
This PR should be fully covered elsewhere. Closing. @dPys I haven't copied your branch to upstream, so please do not remove it until the other PRs are all merged. Thanks for this!

**Pull request closed without merging**